### PR TITLE
Remove broadcasting before ND-batching

### PR DESCRIPTION
**Pull request by abocquet**
Who needs broadcasting anyway ? Remove broadcasting before ND-batching.
Also removes `expand_as_broadcastable`, that isn't used anymore.

Follows https://github.com/dynamiqs/dynamiqs/pull/562

Related to DYN-197.

### CODE REVIEW

File: dynamiqs/mesolve/mesolve.py [hunk 1]
```diff
@@ -133,23 +134,17 @@ def _vectorized_mesolve(
...
-    # this leaf should be vmapped on.
-
-    # the result is vectorized over `_saved` and `infos`
-    out_axes = MEResult(None, None, None, None, 0, 0)
-
     if not options.cartesian_batching:
         broadcast_shape = jnp.broadcast_shapes(
             H.shape[:-2], rho0.shape[:-2], *[jump_op.shape[:-2] for jump_op in jump_ops]
         )
+        rho0 = jnp.broadcast_to(rho0, broadcast_shape + rho0.shape[-2:])
```
Comment (pierreguilmin) [inline dynamiqs/mesolve/mesolve.py:141]:
```suggestion
        broadcast_shape = jnp.broadcast_shapes(H.shape, rho0.shape, *[L.shape for L in jump_ops])
        rho0 = jnp.broadcast_to(rho0, broadcast_shape)
```
Same for sesolve.

Comment (pierreguilmin) [reply dynamiqs/mesolve/mesolve.py:141]: Why do we need to broadcast the initial state?

File: dynamiqs/mesolve/mesolve.py [hunk 2]
```diff
@@ -167,6 +162,8 @@ def broadcast(x: TimeArray) -> TimeArray:
         f = _cartesian_vectorize(_mesolve, n_batch, out_axes)
     else:
         f = _flat_vectorize(_mesolve, n_batch, out_axes)
+        H = squeeze_ones(H)
+        jump_ops = list(map(squeeze_ones, jump_ops))
```
Comment (pierreguilmin) [inline dynamiqs/mesolve/mesolve.py:166]: If the passed Hamiltonian is of shape `(1, n, n)` or `(2, 1, n, n)` we don't want to squeeze the one. IMO `flat_vectorize` the magic you propose should be done by `_flat_vectorize`, it has enough information with the shape of the objects.

File: dynamiqs/time_array.py [hunk 3]
```diff
@@ -222,6 +222,21 @@ def reshape(self, *new_shape: int) -> TimeArray:
...
         """
 
+    def squeeze(self, axis: int) -> TimeArray:
+        """Squeeze a time-array.
+
+        Args:
+            axis: Axis to squeeze.
+
+        Returns:
+        New time-array object with squeezed_shape
```
Comment (pierreguilmin) [inline dynamiqs/time_array.py:232]:
```suggestion
        Returns:
            New time-array object with squeezed shape.
```
If possible it'd be great to render the doc when you modify it, to check that it looks good 😊

**Review by pierreguilmin — CHANGES_REQUESTED** [review]
Nice!

File: dynamiqs/mesolve/mesolve.py (hunk 1 above)
Comment (abocquet) [reply dynamiqs/mesolve/mesolve.py:141]: If your Hamiltonian is (k,n,n) for instance, your state must be (k,n,n) too. Otherwise you cannot store the information of all situations. You can extend this reasoning for additional dimensions.

File: dynamiqs/time_array.py (hunk 3 above)
Comment (abocquet) [reply dynamiqs/time_array.py:232]: Right, sorry I will

File: dynamiqs/mesolve/mesolve.py (hunk 2 above)
Comment (abocquet) [reply dynamiqs/mesolve/mesolve.py:166]: You have to squeeze for emulating the broadcasting. I think it works as is but I'll add a test for your edge case.

File: dynamiqs/mesolve/mesolve.py (hunk 1 above)
Comment (pierreguilmin) [reply dynamiqs/mesolve/mesolve.py:141]: But it's automatic with broadcasting, i.e. the first time you do H @ rho you'll get a result with shape (k, n, n).

Comment (pierreguilmin) [reply dynamiqs/mesolve/mesolve.py:141]: Ha but then maybe JAX/diffrax is happy that rho keeps the same shape all along?

File: dynamiqs/mesolve/mesolve.py (hunk 2 above)
Comment (pierreguilmin) [reply dynamiqs/mesolve/mesolve.py:166]: But you're also going to squeeze the original dimensions equal to one no?

Comment (abocquet) [reply dynamiqs/mesolve/mesolve.py:166]: Rho is not squeezed so shapes may be preserved

File: dynamiqs/mesolve/mesolve.py (hunk 1 above)
Comment (abocquet) [reply dynamiqs/mesolve/mesolve.py:141]: After second though your comment won't hold for sesolve because it would broadcast a bra (n,1) into a dm (n,n)

Comment (pierreguilmin) [reply dynamiqs/mesolve/mesolve.py:141]: Because broadcsasting is from the left, so in the case of sesolve it would be (k, n, 1).

Comment (pierreguilmin) [reply dynamiqs/mesolve/mesolve.py:141]: No, it would be `(k, n, 1)`:
```python
H = jnp.ones((5, 2, 2))
rho = jnp.ones((2, 2))
print((H @ rho).shape)  # (5, 2, 2)
psi = jnp.ones((2, 1))
print((H @ psi).shape)  # (5, 2, 1)
```

**Review by gautierronan — CHANGES_REQUESTED** [review]

File: dynamiqs/mesolve/mesolve.py (hunk 1 above)
Comment (abocquet) [reply dynamiqs/mesolve/mesolve.py:141]: While I do agree in theory, in practise tests break when I remove these lines.

Comment (gautierronan) [reply dynamiqs/mesolve/mesolve.py:140]: I don't have time to investigate, and it's pretty minor, so I'd be tempted to keep it like that and to simply track it in the Linear. Unless @pierreguilmin you have an idea?

File: dynamiqs/mesolve/mesolve.py [hunk 4]
```diff
@@ -167,6 +161,8 @@ def broadcast(x: TimeArray) -> TimeArray:
         f = _cartesian_vectorize(_mesolve, n_batch, out_axes)
     else:
         f = _flat_vectorize(_mesolve, n_batch, out_axes)
+        H = H.squeeze()
+        jump_ops = [x.squeeze() for x in jump_ops]
```
Comment (pierreguilmin) [inline dynamiqs/mesolve/mesolve.py:165]: I don't see how this could work. The output _shape_ will be correct, but the computation will be the same for H `(2, 1, n, n)` and `(1, 2, n, n)` (it will always be squeeze to `(2, n, n)`.

See previous comment
> If the passed Hamiltonian is of shape (1, n, n) or (2, 1, n, n) we don't want to squeeze the one. IMO the magic you propose should be done by _flat_vectorize, it has enough information with the shape of the objects.

**Review by pierreguilmin — CHANGES_REQUESTED** [review]
I think squeezing is not the correct approach, see comment.
